File: app/src/main/java/nl/rsdt/japp/jotial/maps/management/StandardMapItemController.java
package nl.rsdt.japp.jotial.maps.management;

import android.os.Bundle;
import android.os.Parcelable;

import java.util.ArrayList;
import nl.rsdt.japp.jotial.maps.management.transformation.AbstractTransducer;

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 4-9-2016
 * Description...
 */
public abstract class StandardMapItemController<I extends Parcelable, O extends AbstractTransducer.StandardResult<I>> extends MapItemController<ArrayList<I>, O> {

    protected ArrayList<I> items = new ArrayList<>();

    @Override
    public void onIntentCreate(Bundle bundle) {
        super.onIntentCreate(bundle);
        if(bundle != null) {
            O result = bundle.getParcelable(getBundleId());
            if(result != null) {
                this.items = result.getItems();
            }
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        if(savedInstanceState != null)
        {
            if(savedInstanceState.containsKey(getBundleId()))
            {
                this.items = savedInstanceState.getParcelableArrayList(getBundleId());
                O result = getTransducer().generate(items);
                if(googleMap != null) {
                    processResult(result);
                }
                else {
                    buffer = result;
                }
            }
        }
    }

    @Override
    public void onSaveInstanceState(Bundle saveInstanceState) {
        saveInstanceState.putParcelableArrayList(getBundleId(), items);
    }

    @Override
    public void merge(O other) {
        clear();
        processResult(other);
    }

    @Override
    protected void processResult(O result) {
        super.processResult(result);
        this.items.addAll(result.getItems());
    }

    @Override
    protected void clear() {
        super.clear();
        items.clear();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if(items != null)
        {
            /**
             * Do not clear the list or the parcelable list inside the bundle will be empty as well
             * */
            items = null;
        }
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/management/StandardMapItemController.kt
package nl.rsdt.japp.jotial.maps.management

import android.os.Bundle
import android.os.Parcelable
import nl.rsdt.japp.jotial.maps.management.transformation.AbstractTransducer
import nl.rsdt.japp.jotial.maps.wrapper.IJotiMap
import nl.rsdt.japp.jotial.maps.wrapper.osm.OsmJotiMap
import java.util.*

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 4-9-2016
 * Description...
 */
abstract class StandardMapItemController<I : Parcelable, O : AbstractTransducer.StandardResult<I>>(jotiMap: IJotiMap) : MapItemController<ArrayList<I>, O>(jotiMap) {

    protected var items: ArrayList<I> = ArrayList()

    override fun onIntentCreate(bundle: Bundle) {
        super.onIntentCreate(bundle)
        val result = bundle.getParcelable<O>(bundleId)
        if (result != null) {
            this.items = result.items
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        if (savedInstanceState?.containsKey(bundleId)== true) {
            this.items = savedInstanceState.getParcelableArrayList(bundleId)!!
            val result = transducer.generate(items)
            processResult(result)
        }
    }

    override fun onSaveInstanceState(saveInstanceState: Bundle?) {
        saveInstanceState?.putParcelableArrayList(bundleId, items)
    }

    override fun merge(other: O) {
        clear()
        processResult(other)
    }

    override fun processResult(result: O) {
        super.processResult(result)
        this.items.addAll(result.items)
    }

    override fun clear() {
        super.clear()
        items.clear()
    }

    override fun onDestroy() {
        super.onDestroy()
    }
}
